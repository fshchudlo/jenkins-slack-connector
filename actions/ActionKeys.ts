export const ActionKeys = {
    RUN_JENKINS_PIPELINE: "run_jenkins_pipeline",
    RUN_PARAMETERIZED_PIPELINE: "run_parameterized_jenkins_pipeline",
    SUBMIT_PENDING_INPUT: "submit_pending_jenkins_input",
    APORT_PENDING_INPUT: "abort_pending_jenkins_input",
    DISPLAY_CANCELLATION_MESSAGE: "display_cancellation_message",
    SAVE_ACCESS_TOKENS: "save_access_tokens",
    PENDING_INPUT_REQUEST_MESSAGE_PATTERN: ":keyboard:"
};
