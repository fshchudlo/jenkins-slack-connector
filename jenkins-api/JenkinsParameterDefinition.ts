export type JenkinsParameterDefinition = {
    type: "StringParameterDefinition" | "TextParameterDefinition" | "ChoiceParameterDefinition" | "BooleanParameterDefinition";
    name: string;
    description?: string;
    defaultParameterValue?: { value: any };
    definition?: {
        choices?: string[];
        defaultVal?: any;
    };
    choices?: string[];
}
