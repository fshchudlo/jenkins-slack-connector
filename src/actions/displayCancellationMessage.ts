import { AllMiddlewareArgs, SlackActionMiddlewareArgs } from "@slack/bolt";

export async function displayCancellationMessage({ ack, respond }: AllMiddlewareArgs & SlackActionMiddlewareArgs) {
    await ack();

    await respond("Ok, I've canceled the operation. See you soon.");
}
