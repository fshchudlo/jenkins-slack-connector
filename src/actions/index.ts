export * from "./displayCancellationMessage";
export * from "./saveJenkinsAccessToken";

export * from "./jenkins_pipeline_running/runParameterizedJenkinsPipeline";
export * from "./jenkins_pipeline_running/runJenkinsPipelineOrDisplayParametersForm";
export * from "./jenkins_pipeline_running/displayPipelineSelectionForm";

export * from "./jenkins_input_handling/abortJenkinsPendingInput";
export * from "./jenkins_input_handling/displayJenkinsPendingInputForm";
export * from "./jenkins_input_handling/submitJenkinsPendingInput";

export * from "./ActionKeys";
