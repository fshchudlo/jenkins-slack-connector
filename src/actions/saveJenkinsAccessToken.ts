import {getSlackFormValues} from "./helpers/getSlackFormValues";
import {section} from "../building-blocks";
import {AllMiddlewareArgs, SlackViewMiddlewareArgs} from "@slack/bolt";
import { CredentialsStore } from "../CredentialsStore";

export async function saveJenkinsAccessToken({ack, client, body, context}: AllMiddlewareArgs & SlackViewMiddlewareArgs) {
    const respond = async (title: string, ...blocks: any[]): Promise<void> => {
        await ack({
            response_action: "update",
            view: {
                type: "modal",
                title: {type: "plain_text", text: title},
                blocks: blocks
            }
        });
    };

    const payload: any = getSlackFormValues(body.view.state.values);
    const userInfo = (await client.users.info({user: context.userId})).user;

    CredentialsStore.saveCredentials(context.userId, {username: userInfo.profile.email, password: payload.jenkins_token});

    await respond("Success", section(":thumbsup: Token was saved. Now you can use any of my features"));
}