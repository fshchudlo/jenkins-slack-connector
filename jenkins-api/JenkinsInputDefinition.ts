import {JenkinsParameterDefinition} from "./JenkinsParameterDefinition";

export type JenkinsInputDefinition = {
    id: string;
    message: string;
    proceedText: string;
    proceedUrl: string;
    abortUrl: string;
    inputs: JenkinsParameterDefinition[];
}