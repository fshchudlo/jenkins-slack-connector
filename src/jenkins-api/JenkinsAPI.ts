import axios, {AxiosResponse, Method} from "axios";
import {BasicAuthCredentials} from "./BasicAuthCredentials";
import {JenkinsInputDefinition} from "./JenkinsInputDefinition";
import {JenkinsParameterDefinition} from "./JenkinsParameterDefinition";

const JENKINS_API_POSTFIX = "/api/json";

export class JenkinsAPI {
    private readonly auth: BasicAuthCredentials;
    private readonly jenkinsUrl: string;

    constructor(auth: BasicAuthCredentials, jenkinsUrl: string) {
        this.auth = auth;
        this.jenkinsUrl = jenkinsUrl;
    }

    private async request(url: string, method: Method = "GET", params?: any): Promise<AxiosResponse> {
        return axios.request({
            method,
            url,
            params,
            auth: this.auth,
            headers: {"accept": "application/json", "accept-encoding": "gzip, deflate, br"}
        });
    }

    async startJob(jobName: string, jobParams?: any): Promise<string> {
        const runUrlPart = jobParams ? "buildWithParameters" : "build";
        const runJobUrl = `${this.jenkinsUrl}/${jobName}/${runUrlPart}`;

        const queuedJobUrl = (await this.request(runJobUrl, "POST", jobParams)).headers.location;

        // give Jenkins time to start the job
        await new Promise(resolve => setTimeout(resolve, 300));

        const startedJob = (await this.request(`${queuedJobUrl}${JENKINS_API_POSTFIX}`, "GET")).data;
        if (startedJob.blocked) {
            return Promise.reject(startedJob.why);
        }
        return startedJob.executable.url;
    }

    async getJobParameters(jobPath: string): Promise<JenkinsParameterDefinition[] | null> {
        const url = `${this.jenkinsUrl}/${jobPath}/${JENKINS_API_POSTFIX}`;
        const buildInfo = (await this.request(url, "GET")).data;
        const parametersDefinition = buildInfo.property.find((item: object) => "parameterDefinitions" in item);
        return parametersDefinition?.parameterDefinitions || null;
    }

    async getJobPendingInput(jobUrl: string): Promise<JenkinsInputDefinition> {
        const url = `${jobUrl}/wfapi/nextPendingInputAction`;
        return (await this.request(url, "GET")).data;
    }

    async abortJobPendingInput(abortUrl: string): Promise<void> {
        const fullAbortUrl = `${this.jenkinsUrl}/${abortUrl}`;
        await this.request(fullAbortUrl, "POST");
    }

    async submitJobPendingInput(buildUrl: string, submitUrl: string, formValues: any): Promise<void> {
        const inputDefinition = await this.getJobPendingInput(buildUrl);

        if (!inputDefinition || !submitUrl.includes(`inputSubmit?inputId=${inputDefinition.id}`)) {
            throw new Error("Input was already submitted or timed out");
        }
        this.appendUnselectedCheckboxes(formValues, inputDefinition);

        const convertedValues = Object.entries(formValues).map(([key, value]) => ({name: key, value}));
        await this.request(`${this.jenkinsUrl}${submitUrl}`, "POST", {json: JSON.stringify({parameter: convertedValues})});
    }

    private appendUnselectedCheckboxes(formValues: any, inputDefinition: JenkinsInputDefinition): Record<string, any> {
        inputDefinition.inputs.forEach(parameter => {
            if (parameter.type === "BooleanParameterDefinition" && !(parameter.name in formValues)) {
                formValues[parameter.name] = false;
            }
        });
        return formValues;
    }
}
