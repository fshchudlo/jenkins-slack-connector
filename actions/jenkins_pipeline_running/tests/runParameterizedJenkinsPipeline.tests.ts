import {runParameterizedJenkinsPipeline} from "../../../actions";
import {JenkinsAPI} from "../../../jenkins-api/JenkinsAPI";

jest.mock("../../../jenkins-api/JenkinsAPI");

describe("runParameterizedJenkinsPipeline", () => {
    let context: any;
    let body: any;
    let payload: any;
    let ack: jest.Mock;
    let respond: jest.Mock;

    beforeEach(() => {
        context = {jenkinsAPI: new JenkinsAPI(null, null)};
        body = {
            state: {
                values: {
                    "someStringParam": {
                        "someStringParam": {
                            "type": "plain_text_input",
                            "value": "test"
                        }
                    }
                }
            }
        };
        payload = {block_id: "pipelineId", value: "Pipeline Display Name"};
        ack = jest.fn();
        respond = jest.fn();
    });

    it("runs the parameterized pipeline", async () => {
        jest.spyOn(context.jenkinsAPI, "startJob").mockResolvedValueOnce("http://example.com/job");

        await runParameterizedJenkinsPipeline({context, body, payload, ack, respond});

        expect(context.jenkinsAPI.startJob).toHaveBeenCalledWith("pipelineId", {someStringParam: "test"});
        expect(respond).toHaveBeenCalled();
        expect(respond.mock.calls[0][0]).toMatchSnapshot();
    });

    it("runs the parameterized pipeline with no values", async () => {
        body.state.values = {}
        jest.spyOn(context.jenkinsAPI, "startJob").mockResolvedValueOnce("http://example.com/job");

        await runParameterizedJenkinsPipeline({context, body, payload, ack, respond});

        expect(context.jenkinsAPI.startJob).toHaveBeenCalledWith("pipelineId", {});
        expect(respond).toHaveBeenCalled();
        expect(respond.mock.calls[0][0]).toMatchSnapshot();
    });

    it("responds with an error message if running the parameterized pipeline fails", async () => {
        const error = new Error("Network error");
        jest.spyOn(context.jenkinsAPI, "startJob").mockRejectedValueOnce(error);

        await runParameterizedJenkinsPipeline({context, body, payload, ack, respond});

        expect(context.jenkinsAPI.startJob).toHaveBeenCalledWith("pipelineId", {someStringParam: "test"});
        expect(respond).toHaveBeenCalled();
        expect(respond.mock.calls[0][0]).toMatchSnapshot();
    });
});