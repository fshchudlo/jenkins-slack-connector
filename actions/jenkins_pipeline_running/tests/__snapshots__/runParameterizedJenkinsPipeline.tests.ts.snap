// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runParameterizedJenkinsPipeline responds with an error message if running the parameterized pipeline fails 1`] = `
{
  "blocks": [
    {
      "text": {
        "text": ":octagonal_sign: *Unable to run pipeline* 
\`Error: Network error\`",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
}
`;

exports[`runParameterizedJenkinsPipeline runs the parameterized pipeline 1`] = `
{
  "blocks": [
    {
      "text": {
        "text": ":rocket: Here is the <http://example.com/job|Pipeline Display Name> pipeline that you run.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "elements": [
        {
          "text": "Submitted values:",
          "type": "plain_text",
        },
      ],
      "type": "context",
    },
    {
      "elements": [
        {
          "text": "	â€¢ someStringParam: test",
          "type": "plain_text",
        },
      ],
      "type": "context",
    },
  ],
}
`;

exports[`runParameterizedJenkinsPipeline runs the parameterized pipeline with no values 1`] = `
{
  "blocks": [
    {
      "text": {
        "text": ":rocket: Here is the <http://example.com/job|Pipeline Display Name> pipeline that you run.",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    {
      "elements": [
        {
          "text": "Submitted values:",
          "type": "plain_text",
        },
      ],
      "type": "context",
    },
  ],
}
`;
